library/src/ipfs_client/unix_fs/dir_shard.cc

```diff
@@ -0,0 +1,145 @@
+#include "dir_shard.h"
+
+#include <ipfs_client/block_storage.h>
+#include <ipfs_client/unixfs_path_resolver.h>
+
+#include <smhasher/MurmurHash3.h>
+
+#include <vocab/endian.h>
+
+#include "log_macros.h"
+
+#include <absl/strings/match.h>
+
+using Self = ipfs::unix_fs::DirShard;
+
+bool Self::Process(std::unique_ptr<NodeHelper>& next_helper,
+                   std::shared_ptr<DagListener> listener,
+                   std::function<void(std::string, Priority)> requestor,
+                   std::string& target_cid) {
+  if (!block()) {
+    requestor(cid_, resolver_->priority());
+    return false;
+  }
+  SomePrefetch(requestor);
+  auto& block = *(this->block());
+  // node.Data.hashType indicates a multihash function to use to digest path
+  // components used for sharding.
+  //  It MUST be murmur3-x64-64 (multihash 0x22).
+  GOOGLE_DCHECK_EQ(block.fsdata().hashtype(), 0x22);
+  if (next_path_element_.empty()) {
+    // TODO - there could one day be a HAMT directory with no index.html
+    next_path_element_ = "index.html";
+    hamt_hexs_.clear();
+  }
+  /* stolen from spec
+   * ####### Path resolution on HAMTs
+   * Steps:
+   * 1. Take the current path component...
+   */
+  bool missing_descendents = false;
+  if (missing_descendents) {
+    LOG(WARNING)
+        << "Waiting on more blocks before dealing with this HAMT node.";
+    return false;
+  }
+  L_VAR(next_path_element_)
+  auto fanout = block.fsdata().has_fanout() ? block.fsdata().fanout() : 256;
+  if (hamt_hexs_.empty()) {
+    LOG(INFO) << "Had no hexes, hash path element: " << next_path_element_;
+    HashPathElement(fanout);
+  }
+  //  for (auto& h : hamt_hexs_) {
+  //    LOG(INFO) << "Hex: " << h;
+  //  }
+  bool found = false;
+  block.List([&](auto& name, auto cid) {
+    // Fun fact: there is a spec-defined sort order to these children.
+    // We *could* do a binary search.
+    if (!absl::StartsWith(name, hamt_hexs_.front())) {
+      //        LOG(INFO) << name << " isn't the right child for " <<
+      //        hamt_hexs_.front() ;
+      return true;
+    }
+    found = true;
+    LOG(INFO) << "As we move down a Hamt, " << cid_ << " -> " << name << " / "
+              << cid << " fanout=" << block.fsdata().fanout();
+    target_cid = cid;
+    if (absl::EndsWith(name, next_path_element_)) {
+      LOG(INFO) << "Found our calling! Leaving the Hamt in favor of " << name;
+      next_helper.reset();  // Let higher detect what the next level down is
+    } else if (hamt_hexs_.front() == name) {
+      LOG(INFO) << "One more level down the Hamt, following " << name;
+      auto help = std::make_unique<DirShard>(next_path_element_);
+      Delegate(*help);
+      help->cid(cid);
+      help->hamt_hexs_.assign(std::next(hamt_hexs_.begin()), hamt_hexs_.end());
```

Comment:
At one point in this refactor I had absent-mindedly omitted the initial `help->` from this line.

The result was one of the more amusingly-symptomed bugs I've seen recently. I tried to load the landing page of IPFS Wikipedia and instead it loaded "Tonight, We Ride"